### DISCUSSION TITLE
AttributeError: 'WindowsPath' object has no attribute 'set_alpn_protocols'

### DISCUSSION THREAD

**Discussion by PabloRMira** [encode/httpcore#984] [Potential Issue]
Using the openai package I get an APIConnectionError with direct cause

AttributeError: 'WindowsPath' object has no attribute 'set_alpn_protocols'

which lies within httpcore. Here is the relevant traceback:

```
File c:\DEV\Miniconda3\envs\myenv\Lib\site-packages\httpcore\_sync\http_proxy.py:308, in TunnelHTTPConnection.handle_request(self, request)
    307 alpn_protocols = ["http/1.1", "h2"] if self._http2 else ["http/1.1"]
--> 308 ssl_context.set_alpn_protocols(alpn_protocols)
    310 kwargs = {
    311     "ssl_context": ssl_context,
    312     "server_hostname": self._remote_origin.host.decode("ascii"),
    313     "timeout": timeout,
    314 }

AttributeError: 'WindowsPath' object has no attribute 'set_alpn_protocols'
```

httpcore version = '1.0.7'

Could you please help me fix this issue? Thanks in advance!

**Comment by TayfurYldz** [comment]
That traceback means httpcore received a `pathlib.WindowsPath` where it requires an `ssl.SSLContext`.

At the failing line, httpcore does this conceptually:

```python
ssl_context = default_ssl_context() if self._ssl_context is None else self._ssl_context
ssl_context.set_alpn_protocols(...)
```

So if the object is a `WindowsPath`, the problem isn't ALPN support on Windows; something upstream supplied a CA/certificate *path* as the HTTP client's SSL context instead of constructing an SSL context from that path.

The first useful check is to inspect how the OpenAI client / underlying HTTPX client is configured, particularly any custom `verify=`, transport, proxy, or SSL setting. A filesystem CA bundle path should be converted to a context, e.g.:

```python
import ssl

ctx = ssl.create_default_context(cafile=str(ca_bundle_path))
```

and that context is what the transport-level `ssl_context` expects.

If you didn't configure SSL explicitly, also check environment-driven CA settings and the exact versions of `openai`, `httpx`, and `httpcore`, because the bad type may have been introduced in the layer that constructs httpcore's connection pool.

I checked current httpcore as well: the sync proxy connection's `ssl_context` parameter is typed as `ssl.SSLContext | None`, and immediately before your failing call it either creates the default context or uses that supplied object unchanged. So changing `set_alpn_protocols` in httpcore would only mask the actual type/configuration error; the fix is to ensure the transport receives an `SSLContext`, not a `Path`.
